### PR TITLE
Blazor API Review: RenderTree types

### PR DESCRIPTION
**Pull request by rynowak**
Fixes: #12243 

This change adds an analyzer to scare you away from `.RenderTree` as well as some S P O O K Y warnings in the comments of the classes.

### CODE REVIEW

File: src/Components/Analyzers/src/Resources.resx
Comment (SteveSandersonMS) [inline src/Components/Analyzers/src/Resources.resx]:
```suggestion
    <value>The type or member {0} is is not recommended for use outside of the Blazor framework. Types defined in 'Microsoft.AspNetCore.Components.RenderTree' will change in future releases.</value>
```

I know we *hope* not to change it outside major releases, but we don't have to give a written promise here :)

File: src/Components/Analyzers/src/ComponentInternalUsageDiagnosticAnalzyer.cs
Comment (SteveSandersonMS) [inline src/Components/Analyzers/src/ComponentInternalUsageDiagnosticAnalzyer.cs]: I'm unsure whether we specifically stated which types to consider pubternal. I was expecting this to include `RenderTreeFrame`, `RenderTreeEdit`, `RenderTreeDiff`, or anything else that you would only use when building a custom renderer.

I *didn't* expect us to be forbidding use of `RenderTreeBuilder`, which people would need to do if implementing `IComponent` directly, or writing custom builder logic which I know many component vendors are doing. Should we add a specific exclusion for RTB or maybe move it to a different namespace?

(You previously suggested renaming the `.RenderTree` namespace to `.Rendering` which sounds good to me)

File: src/Components/Analyzers/src/InternalUsageAnalyzer.cs
Comment (SteveSandersonMS) [inline src/Components/Analyzers/src/InternalUsageAnalyzer.cs]: I find it pretty amazing that (or *if*) this sort of thing doesn't impact compile times noticeably. Is it basically because the normal C# compilation process already looks like this internally but is 1000x bigger?

File: src/Components/Analyzers/test/TestFiles/ComponentInternalUsageDiagnoticsAnalyzerTest/UsesRenderTreeFrameAsParameter.cs
Comment (SteveSandersonMS) [inline src/Components/Analyzers/test/TestFiles/ComponentInternalUsageDiagnoticsAnalyzerTest/UsesRenderTreeFrameAsParameter.cs]: What's `/*MM*/`? Is this just adding a comment node to show the analyzer can handle it?

File: src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs]: @JamesNK FYI

File: src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs]:
```suggestion
    /// of the Blazor framework. These types will change in a future release.
```

File: src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs]: Was it intentional to change this line?

Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs]: Well I guess so, since you did the same below. Not sure why though.

File: src/Components/Components/src/RenderTree/ArrayRange.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/ArrayRange.cs]:
```suggestion
    /// of the Blazor framework. These types will change in a future release.
```

File: src/Components/Components/src/RenderTree/RenderTreeDiff.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/RenderTreeDiff.cs]:
```suggestion
    /// of the Blazor framework. These types will change in a future release.
```

File: src/Components/Components/src/RenderTree/RenderTreeEdit.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/RenderTreeEdit.cs]:
```suggestion
    /// of the Blazor framework. These types will change in a future release.
```

File: src/Components/Components/src/RenderTree/RenderTreeEditType.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/RenderTreeEditType.cs]:
```suggestion
    /// of the Blazor framework. These types will change in a future release.
```

File: src/Components/Components/src/RenderTree/RenderTreeFrame.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/RenderTreeFrame.cs]:
```suggestion
    /// of the Blazor framework. These types will change in a future release.
```

File: src/Components/Components/src/RenderTree/RenderTreeFrameType.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/RenderTreeFrameType.cs]:
```suggestion
    /// of the Blazor framework. These types will change in a future release.
```

**Review by SteveSandersonMS — COMMENTED** [review]
My only real concern here is about whether we're pubternalling `RenderTreeBuilder` or not - let's discuss. Besides that this looks good to me!

File: src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs
Comment (JamesNK) [reply src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs]: image

File: src/Components/Analyzers/src/Resources.resx
Comment (rynowak) [reply src/Components/Analyzers/src/Resources.resx]: heh

File: src/Components/Analyzers/src/ComponentInternalUsageDiagnosticAnalzyer.cs
Comment (rynowak) [reply src/Components/Analyzers/src/ComponentInternalUsageDiagnosticAnalzyer.cs]: As previously discussed, the plan to to move `RenderTreeBuilder` to another namespace.

File: src/Components/Analyzers/src/Resources.resx
Comment (rynowak) [reply src/Components/Analyzers/src/Resources.resx]: I'm ok with changing the text. I also don't think we'll be allowed to change this in a minor release.

File: src/Components/Analyzers/src/InternalUsageAnalyzer.cs
Comment (rynowak) [reply src/Components/Analyzers/src/InternalUsageAnalyzer.cs]: The compiler is already pretty smart, and all of the data we're accessing is cached and used already. Basically we register callbacks and then operate on the nodes the compiler gives us. We don't really get to decide what nodes the compiler produces which is why it's efficient.

Here's some data from the MVC template that has more analyzers: `dotnet msbuild /t:rebuild /p:ReportAnalyzer=true /bl`

```txt
Using shared compilation with compiler from directory: C:\Program Files\dotnet\sdk\3.0.100-preview8-013528\Roslyn\bincore
Microsoft (R) Visual C# Compiler version 3.3.0-beta2-19376-02 (fc001df5)
Copyright (C) Microsoft Corporation. All rights reserved.
Total analyzer execution time: 0.088 seconds.
NOTE: Elapsed time may be less than analyzer execution time because analyzers can run concurrently.
Time (s)    %   Analyzer
0.080   91   Microsoft.AspNetCore.Analyzers, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
0.080   91      Microsoft.AspNetCore.Analyzers.StartupAnalzyer
0.005    6   Microsoft.AspNetCore.Mvc.Analyzers, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
0.002    1      Microsoft.AspNetCore.Mvc.Analyzers.TagHelpersInCodeBlocksAnalyzer
0.001    1      Microsoft.AspNetCore.Mvc.Analyzers.TopLevelParameterNameAnalyzer
0.001    1      Microsoft.AspNetCore.Mvc.Analyzers.AvoidHtmlPartialAnalyzer
0.001    1      Microsoft.AspNetCore.Mvc.Analyzers.AttributesShouldNotBeAppliedToPageModelAnalyzer
0.002    2   Microsoft.AspNetCore.Components.Analyzers, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
0.002    1      Microsoft.AspNetCore.Components.Analyzers.ComponentParameterAnalyzer
<0.001   <1      Microsoft.AspNetCore.Components.Analyzers.ComponentParameterUsageAnalyzer
```

The analyzer runtime was less than 9ms. 

Obviously this doesn't include the analyzer getting worked on here, but these analyzers do similar things.

--- 

Ultimately I'd want to make some improvements to this code overall. The approach us using Syntax + Semantic model is the *old bad way* of doing things, and they now have something purpose built for what we need.

File: src/Components/Analyzers/test/TestFiles/ComponentInternalUsageDiagnoticsAnalyzerTest/UsesRenderTreeFrameAsParameter.cs
Comment (rynowak) [reply src/Components/Analyzers/test/TestFiles/ComponentInternalUsageDiagnoticsAnalyzerTest/UsesRenderTreeFrameAsParameter.cs]: This is a marker so the test framework can compare against the location of the diagnostic.

File: src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs
Comment (rynowak) [reply src/Components/Components/src/RenderTree/ArrayBuilderSegment.cs]: Yes. The point is to just make the documentation blank other than the stern warning about breaking changes. It's wierd to fill in the summary with boilerplate but then provide detail in the other thigns. 

In other cases, I preserved existing documentation *as comments* because it has some value to us.

**Comment by rynowak** [comment]
@SteveSandersonMS - any other feedback? Moving the namespace of `RenderTreeBuilder` is part of this PR - https://github.com/aspnet/AspNetCore/pull/12854

File: src/Components/Analyzers/src/ComponentInternalUsageDiagnosticAnalzyer.cs
Comment (SteveSandersonMS) [reply src/Components/Analyzers/src/ComponentInternalUsageDiagnosticAnalzyer.cs]: Ok good.

**Review by SteveSandersonMS — APPROVED** [review]

**Comment by rynowak** [comment]
> Yo, check it, out, this fails on helix due to \r\nvs\n. 

So what's the solution to that? I thought all of these tests were disabled on helix.

**Comment by rynowak** [comment]
FYI this needs to wait until the `RenderTreeBuilder` change is merged. The templates build is failing because the analyzer is doing its job.

**Comment by BrennanConroy** [comment]
@rynowak I'm assuming you didn't change 204 files in this PR 😆

**Comment by rynowak** [comment]
I just merged this bad boy: https://github.com/aspnet/AspNetCore-Tooling/pull/873 which is 623 files on a **SUNDAY**. What makes you think I won't change 204 files?

**Comment by BrennanConroy** [comment]
Because half the commits weren't yours!

**Pull request merged**
